### PR TITLE
kernel: provide Platform abstract interface for device drivers which are designed to works seamlessly on PCI bus, ACPI and DTS

### PR DESCRIPTION
**Pull request by najumon1980**
Provide Platform abstract interface for device drivers which are designed to work on multiple bus topologies such as PCI, ACPI or just DTS base. Driver developers can use the new PLATFORM_DT_DEV_INST_DEFINE macro instead of DEVICE_DT_INST_DEFINE in case they have to design drivers which need to works seamlessly on PCI bus, ACPI and DTS base.  Currently  there is no support for MSI and multi vectors. Following are some of the key features/highlight:

- Abstract underlying bus/parent complexity and providing a generic interface to driver developers. Using the new interface, driver developers need not want to bother about underlying bus interface and associated complexity. 
- Based on their parent node type, the interface will automatically enumerate resource such as MMIO and IRQ resources and configure them which will be completely transparent to device drivers
- The interface will take care of memory map of MMIO address in case platform support MMU
- The new PLATFORM_DT_DEV_INST_DEFINE  interface take ISR as on of the argument and will take care the ISR registration based on underlying bus/parent node  type(PCI/ACPI or just DTS). 
- The new interface add very little overhead (almost NIL) 
 
This PR is depend on PR: https://github.com/zephyrproject-rtos/zephyr/pull/62694

### CODE REVIEW

File: dts/x86/intel/raptor_lake.dtsi [hunk 1]
```diff
@@ -44,6 +45,25 @@
 		#address-cells = <1>;
 	};
 
+	acpi0: acpi0 {
```
Comment (tbursztyka) [inline dts/x86/intel/raptor_lake.dtsi:48]: ACPI is not a hw bus. This is going against DTS idea.

I think we already discussed that.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: dts/x86/intel/raptor_lake.dtsi (hunk 1 above)
Comment (najumon1980) [reply dts/x86/intel/raptor_lake.dtsi]: removed ACPI node. In order to distinguish whether a driver use ACPI interface or not, one can use DT_NODE_HAS_PROP(n, acpi_uid)). 

@galak could you please have a look and share your though ?

File: dts/bindings/acpi/acpi.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# Common fields for PCIe devices
+
+properties:
+  acpi-hw-id:
+    type: string
+    description: Used to supply OSPM with the device’s PNP ID or ACPI ID
+
+  acpi-uid:
+    type: string
+    description: Provides OSPM with a logical device ID that does not change
```
Comment (mbolivar-ampere) [inline dts/bindings/acpi/acpi.yaml:13]: https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#descriptions

File: include/zephyr/dt-bindings/acpi/acpi.h [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define ACPI_IRQ_DETECT 0xFFFFFFFU
```
Comment (mbolivar-ampere) [inline include/zephyr/dt-bindings/acpi/acpi.h:7]: Needs include guard

File: include/zephyr/acpi/acpi.h [hunk 4]
```diff
@@ -88,6 +88,7 @@ struct acpi_mmio_resource {
  */
 #define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, hardware_id, "INVALD")
 
+
```
Comment (mbolivar-ampere) [inline include/zephyr/acpi/acpi.h:91]: Something is wrong with this commit. It isn't doing anything related to the commit description.

**Review by mbolivar-ampere — CHANGES_REQUESTED** [review]
I don't know ACPI and don't have any general objections. Please fix a few minor things.

File: include/zephyr/dt-bindings/acpi/acpi.h (hunk 3 above)
Comment (najumon1980) [reply include/zephyr/dt-bindings/acpi/acpi.h:10]: updated

File: include/zephyr/acpi/acpi.h (hunk 4 above)
Comment (najumon1980) [reply include/zephyr/acpi/acpi.h]: this is due to dependency PR https://github.com/zephyrproject-rtos/zephyr/pull/62694 and rebased it

File: dts/bindings/acpi/acpi.yaml (hunk 2 above)
Comment (najumon1980) [reply dts/bindings/acpi/acpi.yaml]: updated

File: boards/x86/qemu_x86/qemu_x86_64_acpi.dts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+	pcie0: pcie0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "intel,pcie";
+		acpi-hid = "PNP0A08";
+		acpi-uid = "0";
+		ranges;
+
+	smbus0: smbus0 {
```
Comment (finikorg) [inline boards/x86/qemu_x86/qemu_x86_64_acpi.dts:145]: wrong indentation for smbus

File: include/zephyr/arch/x86/x86_acpi_osal.h [hunk 6]
```diff
@@ -26,4 +26,9 @@ static inline void *acpi_rsdp_get(void)
 	return bios_acpi_rsdp_get();
 }
 #endif /* CONFIG_X86_EFI */
+
```
Comment (finikorg) [inline include/zephyr/arch/x86/x86_acpi_osal.h:29]: The commit has syntax error `sperate` -> `separate`

File: lib/acpi/acpi.c [hunk 7]
```diff
@@ -217,6 +217,9 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 	}
 
 	child_dev = (struct acpi_dev *)&acpi.child_dev[acpi.num_dev++];
+	if (!dev_info->HardwareId.String) {
+		goto exit;
+	}
```
Comment (finikorg) [inline lib/acpi/acpi.c:222]: This check should be put after function which get device_info.

File: lib/acpi/acpi_shell.c [hunk 8]
```diff
@@ -220,19 +226,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
...
+	if (!dev) {
+		shell_error(sh, "acpi get device failed for HID: %s", argv[1]);
+		return -EIO;
+	}
+
+	if (dev->path) {
+		shell_print(sh, "Device Path: %s", dev->path);
+
+		if (!acpi_device_mmio_get(dev, &mmio_res)) {
+
```
Comment (finikorg) [inline lib/acpi/acpi_shell.c:297]: No need for empty line here

File: lib/acpi/acpi_shell.c [hunk 9]
```diff
@@ -220,19 +226,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
...
+			for (int i = 0; i < ACPI_RESOURCE_COUNT_GET(&mmio_res); i++) {
+				shell_print(sh, "\tType: %x, Address: %p, Size: %d",
+					ACPI_MULTI_RESOURCE_TYPE_GET(&mmio_res, i),
+					(void *)ACPI_MULTI_MMIO_GET(&mmio_res, i),
+						ACPI_MULTI_RESOURCE_SIZE_GET(&mmio_res, i));
+			}
+		}
+
+		if (!acpi_device_irq_get(dev, &irq_res)) {
+
```
Comment (finikorg) [inline lib/acpi/acpi_shell.c:308]: No need for empty line here

File: drivers/rtc/rtc_mc146818.c [hunk 10]
```diff
@@ -15,9 +15,10 @@
 #include <zephyr/devicetree.h>
 #include <zephyr/drivers/rtc.h>
 #include <zephyr/sys/sys_io.h>
+#include <zephyr/platform.h>
 
-#define RTC_STD_INDEX (DT_INST_REG_ADDR_BY_IDX(0, 0))
-#define RTC_STD_TARGET (DT_INST_REG_ADDR_BY_IDX(0, 1))
+#define RTC_STD_INDEX(dev)  PLATFORM_MMIO_GET(dev)
+#define RTC_STD_TARGET(dev) (PLATFORM_MMIO_GET(dev) + 1)
```
Comment (finikorg) [inline drivers/rtc/rtc_mc146818.c:21]: Is there better way?

File: boards/x86/qemu_x86/qemu_x86_64_acpi.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+identifier: qemu_x86_64
+name: QEMU Emulation for X86_64
+type: qemu
+arch: x86
+toolchain:
+  - zephyr
+  - xtools
+simulation: qemu
+supported:
+  - acpi
```
Comment (finikorg) [inline boards/x86/qemu_x86/qemu_x86_64_acpi.yaml:10]: The board `qemu_x86_64` does also support ACPI, what is the difference then?

**Pull request closed without merging**